File: agentic-adoption-scan/scanner.go
package main

import (
	"fmt"
	"log"
	"strings"
	"time"
)

// ScanResult represents one observation in the tidy data output.
type ScanResult struct {
	ScanTimestamp  string // ISO 8601
	Org            string
	Repo           string
	RepoVisibility string
	RepoLanguage   string
	RepoPushedAt   string
	Category       string
	Indicator      string
	Found          bool
	FilePath       string
	Details        string
}

// Scanner orchestrates the scanning process.
type Scanner struct {
	Client          *GitHubClient
	Cache           *Cache
	Org             string
	Indicators      []Indicator
	ActiveSince     time.Time
	IncludeArchived bool
	Force           bool
	Logger          *log.Logger
}

// Scan runs the full scan and returns results.
func (s *Scanner) Scan() ([]ScanResult, error) {
	repos, err := s.Client.ListOrgRepos(s.Org)
	if err != nil {
		return nil, fmt.Errorf("listing repos: %w", err)
	}

	filtered := s.filterRepos(repos)
	s.Logger.Printf("Scanning %d repos (filtered from %d total)", len(filtered), len(repos))

	now := time.Now().UTC().Format(time.RFC3339)
	var results []ScanResult

	for i, repo := range filtered {
		s.Logger.Printf("[%d/%d] Scanning %s", i+1, len(filtered), repo.FullName)

		// Check cache
		if !s.Force && s.Cache.IsRepoFresh(s.Org, repo.Name, repo.PushedAt) {
			s.Logger.Printf("  Cache hit for %s (pushed_at unchanged)", repo.Name)
			cached := s.Cache.GetRepoResults(s.Org, repo.Name)
			for _, cr := range cached {
				results = append(results, ScanResult{
					ScanTimestamp:  cr.ScannedAt,
					Org:            s.Org,
					Repo:           repo.Name,
					RepoVisibility: repo.Visibility,
					RepoLanguage:   repo.Language,
					RepoPushedAt:   repo.PushedAt,
					Category:       cr.Category,
					Indicator:      cr.Indicator,
					Found:          cr.Found,
					FilePath:       cr.FilePath,
					Details:        cr.Details,
				})
			}
			continue
		}

		repoResults := s.scanRepo(repo, now)
		results = append(results, repoResults...)

		// Update cache
		var cacheResults []CachedIndicator
		for _, r := range repoResults {
			cacheResults = append(cacheResults, CachedIndicator{
				Category:  r.Category,
				Indicator: r.Indicator,
				Found:     r.Found,
				FilePath:  r.FilePath,
				Details:   r.Details,
				ScannedAt: r.ScanTimestamp,
			})
		}
		s.Cache.SetRepoResults(s.Org, repo.Name, repo.PushedAt, cacheResults)
	}

	return results, nil
}

func (s *Scanner) filterRepos(repos []Repo) []Repo {
	var filtered []Repo
	for _, r := range repos {
		if r.Archived && !s.IncludeArchived {
			continue
		}
		if r.PushedAt != "" {
			pushed, err := time.Parse(time.RFC3339, r.PushedAt)
			if err == nil && pushed.Before(s.ActiveSince) {
				continue
			}
		}
		filtered = append(filtered, r)
	}
	return filtered
}

func (s *Scanner) scanRepo(repo Repo, timestamp string) []ScanResult {
	owner := s.Org
	var results []ScanResult

	for _, ind := range s.Indicators {
		result := ScanResult{
			ScanTimestamp:  timestamp,
			Org:            s.Org,
			Repo:           repo.Name,
			RepoVisibility: repo.Visibility,
			RepoLanguage:   repo.Language,
			RepoPushedAt:   repo.PushedAt,
			Category:       ind.Category,
			Indicator:      ind.Name,
		}

		switch ind.SearchType {
		case FileExists:
			exists, _, err := s.Client.CheckPathExists(owner, repo.Name, ind.Target)
			if err != nil {
				s.Logger.Printf("  Warning: error checking %s in %s: %v", ind.Target, repo.Name, err)
			}
			result.Found = exists
			if exists {
				result.FilePath = ind.Target
			}

		case DirectoryExists:
			exists, isDir, err := s.Client.CheckPathExists(owner, repo.Name, ind.Target)
			if err != nil {
				s.Logger.Printf("  Warning: error checking %s in %s: %v", ind.Target, repo.Name, err)
			}
			result.Found = exists && isDir
			if result.Found {
				result.FilePath = ind.Target
			}

		case ContentSearch:
			sr, err := s.Client.SearchCode(owner, repo.Name, ind.Target)
			if err != nil {
				s.Logger.Printf("  Warning: error searching %s in %s: %v", ind.Target, repo.Name, err)
			} else if sr.TotalCount > 0 {
				result.Found = true
				var paths []string
				for _, item := range sr.Items {
					paths = append(paths, item.Path)
				}
				result.FilePath = strings.Join(paths, "; ")
				result.Details = fmt.Sprintf("%d matches", sr.TotalCount)
			}

		case WorkflowSearch:
			sr, err := s.Client.SearchCodeInWorkflows(owner, repo.Name, ind.Target)
			if err != nil {
				s.Logger.Printf("  Warning: error searching workflows in %s: %v", repo.Name, err)
			} else if sr.TotalCount > 0 {
				result.Found = true
				var paths []string
				for _, item := range sr.Items {
					paths = append(paths, item.Path)
				}
				result.FilePath = strings.Join(paths, "; ")
				result.Details = fmt.Sprintf("%d workflow files", sr.TotalCount)
			}
		}

		if result.Found {
			s.Logger.Printf("  Found: %s/%s", ind.Category, ind.Name)
		}
		results = append(results, result)
	}

	return results
}

File: agentic_adoption_scan/scanner.py
"""Scanner: orchestrates scanning a GitHub org for agentic adoption indicators.

Port of agentic-adoption-scan/scanner.go.
"""

from __future__ import annotations

import logging
from datetime import datetime, timezone
from typing import Optional

from agentic_adoption_scan.cache import Cache, CachedIndicator
from agentic_adoption_scan.github import GitHubClient, Repo
from agentic_adoption_scan.indicators import Indicator, SearchType
from agentic_adoption_scan.models import ScanResult

logger = logging.getLogger(__name__)


class Scanner:
    """Orchestrates the full org scan.

    Parameters
    ----------
    client:
        Authenticated GitHub client.
    cache:
        Persistent result cache.
    org:
        GitHub organisation slug to scan.
    indicators:
        List of indicators to check per repo.
    active_since:
        Exclude repos whose last push is before this datetime.
    include_archived:
        When False (default), skip archived repos.
    force:
        When True, bypass the cache and re-scan every repo.
    log:
        Optional logger; defaults to the module logger.
    """

    def __init__(
        self,
        client: GitHubClient,
        cache: Cache,
        org: str,
        indicators: list[Indicator],
        active_since: datetime,
        include_archived: bool = False,
        force: bool = False,
        log: Optional[logging.Logger] = None,
    ) -> None:
        self.client = client
        self.cache = cache
        self.org = org
        self.indicators = indicators
        self.active_since = active_since
        self.include_archived = include_archived
        self.force = force
        self._log = log or logger

    # ------------------------------------------------------------------
    # Public entry point
    # ------------------------------------------------------------------

    def scan(self) -> list[ScanResult]:
        """Run the full scan and return tidy results."""
        repos = self.client.list_org_repos(self.org)
        filtered = self._filter_repos(repos)
        self._log.info(
            "Scanning %d repos (filtered from %d total)", len(filtered), len(repos)
        )

        now = datetime.now(tz=timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
        results: list[ScanResult] = []

        for i, repo in enumerate(filtered, start=1):
            self._log.info("[%d/%d] Scanning %s", i, len(filtered), repo.full_name)

            # Cache hit: reuse stored results
            if not self.force and self.cache.is_repo_fresh(
                self.org, repo.name, repo.pushed_at
            ):
                self._log.info(
                    "  Cache hit for %s (pushed_at unchanged)", repo.name
                )
                cached = self.cache.get_repo_results(self.org, repo.name) or []
                for cr in cached:
                    results.append(
                        ScanResult(
                            scan_timestamp=cr.scanned_at,
                            org=self.org,
                            repo=repo.name,
                            repo_visibility=repo.visibility,
                            repo_language=repo.language,
                            repo_pushed_at=repo.pushed_at,
                            category=cr.category,
                            indicator=cr.indicator,
                            found=cr.found,
                            file_path=cr.file_path,
                            details=cr.details,
                        )
                    )
                continue

            repo_results = self._scan_repo(repo, now)
            results.extend(repo_results)

            # Update cache
            cache_indicators = [
                CachedIndicator(
                    category=r.category,
                    indicator=r.indicator,
                    found=r.found,
                    file_path=r.file_path,
                    details=r.details,
                    scanned_at=r.scan_timestamp,
                )
                for r in repo_results
            ]
            self.cache.set_repo_results(
                self.org, repo.name, repo.pushed_at, cache_indicators
            )

        return results

    # ------------------------------------------------------------------
    # Filtering
    # ------------------------------------------------------------------

    def _filter_repos(self, repos: list[Repo]) -> list[Repo]:
        """Drop archived repos (unless include_archived) and inactive repos."""
        filtered: list[Repo] = []
        for repo in repos:
            if repo.archived and not self.include_archived:
                continue
            if repo.pushed_at:
                try:
                    pushed = datetime.fromisoformat(
                        repo.pushed_at.replace("Z", "+00:00")
                    )
                    # Make active_since offset-aware if it isn't already
                    since = self.active_since
                    if since.tzinfo is None:
                        since = since.replace(tzinfo=timezone.utc)
                    if pushed < since:
                        continue
                except ValueError:
                    pass  # can't parse; include the repo
            filtered.append(repo)
        return filtered

    # ------------------------------------------------------------------
    # Per-repo scanning
    # ------------------------------------------------------------------

    def _scan_repo(self, repo: Repo, timestamp: str) -> list[ScanResult]:
        """Check every indicator against *repo*. Returns one result per indicator."""
        owner = self.org
        results: list[ScanResult] = []

        for ind in self.indicators:
            result = ScanResult(
                scan_timestamp=timestamp,
                org=self.org,
                repo=repo.name,
                repo_visibility=repo.visibility,
                repo_language=repo.language,
                repo_pushed_at=repo.pushed_at,
                category=ind.category,
                indicator=ind.name,
                found=False,
                file_path="",
                details="",
            )

            if ind.search_type == SearchType.FILE_EXISTS:
                try:
                    exists, _ = self.client.check_path_exists(owner, repo.name, ind.target)
                except Exception as exc:  # noqa: BLE001
                    self._log.warning(
                        "  Warning: error checking %s in %s: %s",
                        ind.target,
                        repo.name,
                        exc,
                    )
                    exists = False
                result.found = exists
                if exists:
                    result.file_path = ind.target

            elif ind.search_type == SearchType.DIRECTORY_EXISTS:
                try:
                    exists, is_dir = self.client.check_path_exists(
                        owner, repo.name, ind.target
                    )
                except Exception as exc:  # noqa: BLE001
                    self._log.warning(
                        "  Warning: error checking %s in %s: %s",
                        ind.target,
                        repo.name,
                        exc,
                    )
                    exists, is_dir = False, False
                result.found = exists and is_dir
                if result.found:
                    result.file_path = ind.target

            elif ind.search_type == SearchType.CONTENT_SEARCH:
                try:
                    sr = self.client.search_code(owner, repo.name, ind.target)
                except Exception as exc:  # noqa: BLE001
                    self._log.warning(
                        "  Warning: error searching %s in %s: %s",
                        ind.target,
                        repo.name,
                        exc,
                    )
                    sr = None
                if sr is not None and sr.total_count > 0:
                    result.found = True
                    result.file_path = "; ".join(
                        item["path"] for item in sr.items
                    )
                    result.details = f"{sr.total_count} matches"

            elif ind.search_type == SearchType.WORKFLOW_SEARCH:
                try:
                    sr = self.client.search_code_in_workflows(
                        owner, repo.name, ind.target
                    )
                except Exception as exc:  # noqa: BLE001
                    self._log.warning(
                        "  Warning: error searching workflows in %s: %s",
                        repo.name,
                        exc,
                    )
                    sr = None
                if sr is not None and sr.total_count > 0:
                    result.found = True
                    result.file_path = "; ".join(
                        item["path"] for item in sr.items
                    )
                    result.details = f"{sr.total_count} workflow files"

            if result.found:
                self._log.info("  Found: %s/%s", ind.category, ind.name)

            results.append(result)

        return results
